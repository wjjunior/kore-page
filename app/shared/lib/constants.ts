export const NavItems = {
  INVEST: "invest",
  FAQ: "faq",
  CONTACT: "contact",
  EDUCATION: "education",
};
