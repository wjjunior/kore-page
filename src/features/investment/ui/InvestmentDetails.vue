<template>
  <div class="min-h-screen bg-white pt-[93px] px-40">
    <div class="max-w-7xl mx-auto">
      <div class="flex space-x-6">
        <div class="w-[190px] flex-shrink-0">
          <div class="w-[143px] sticky top-[120px]">
            <nav class="flex flex-col gap-4">
              <a
                v-for="link in navigationLinks"
                :key="link.href"
                :href="link.href"
                class="block text-primary-200 font-hanken-grotesk font-bold text-base leading-5 hover:text-primary-300 transition-colors"
              >
                {{ link.text }}
              </a>
            </nav>
          </div>
        </div>

        <div class="w-[906px] flex-shrink-0">
          <InvestmentCards
            :offering-terms="getOfferingTerms"
            :documents="getDocuments"
            :loading="loading"
          />

          <div class="mt-8">
            <InvestmentTeam />
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { onMounted } from "vue";
import { useInvestmentData } from "../composables";
import { InvestmentCards, InvestmentTeam } from "./";

// Types
interface NavigationLink {
  href: string;
  text: string;
}

const { getOfferingTerms, getDocuments, loadData, loading } =
  useInvestmentData();

const navigationLinks: NavigationLink[] = [
  { href: "#offering-terms", text: "Offering Terms and Documents" },
  { href: "#team", text: "Team" },
  { href: "#marketing-plan", text: "Marketing Plan" },
  { href: "#faq", text: "FAQ" },
];

onMounted(async () => {
  await loadData();
});
</script>
