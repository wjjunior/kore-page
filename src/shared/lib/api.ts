import type { InvestmentBannerData as InvestmentData } from "./types";

export class InvestmentAPI {
  private static instance: InvestmentAPI;

  private constructor() {}

  static getInstance(): InvestmentAPI {
    if (!InvestmentAPI.instance) {
      InvestmentAPI.instance = new InvestmentAPI();
    }
    return InvestmentAPI.instance;
  }

  async fetchInvestmentsData(): Promise<InvestmentData> {
    await new Promise((resolve) => setTimeout(resolve, 1000));

    const mockData: InvestmentData = {
      daysLeft: 213,
      totalInvestors: 157,
      fundingGoal: 250000,
      fundsRaised: 300000,
      minimumInvestment: 100,
      deadline: "Feb, 2025",
      typeOfSecurity: "Revenue Share Agreement",
      revenueShareDuration: 36,
      categories: ["Fintech", "Investments"],
      companyName: "Kore",
      companyDescription: "Lorem ipsum dolor sit",
      website: "https://site.com",
      offeringTerms: [
        {
          label: "Regulation",
          value: "Regulation Crowdfunding (RegCF)",
        },
        {
          label: "Offering Type",
          value: "Revenue Sharing Agreement",
        },
        {
          label: "Security Type",
          value: "Debt",
        },
        {
          label: "Target Offering",
          value: "$250,000",
        },
        {
          label: "Max Offering",
          value: "$2,000,000",
        },
        {
          label: "Min Investment",
          value: "$100",
        },
        {
          label: "Max Investment",
          value: "$50,000",
        },
        {
          label: "Minimum Hold Period",
          value: "36 months",
        },
        {
          label: "Closing Date",
          value: "Feb 28, 2025 12:59 AM GMT-3",
        },
      ],
      documents: [
        {
          id: 1,
          title: "Form C",
          filename: "FileName_GoesHere.pdf",
        },
        {
          id: 2,
          title: "Custodian and Voting Agreement",
          filename: "FileName_GoesHere.pdf",
        },
        {
          id: 3,
          title: "Future Proof Convertible Note",
          filename: "FileName_GoesHere.pdf",
        },
        {
          id: 4,
          title: "Future Proof Convertible Note",
          filename: "FileName_GoesHere.pdf",
        },
        {
          id: 5,
          title: "Future Proof Convertible Note",
          filename: "FileName_GoesHere.pdf",
        },
      ],
      teamMembers: [
        {
          id: 1,
          name: "Jane Smith",
          position: "Director of Product Marketing",
          image:
            "https://i.postimg.cc/KvMRHGh4/73350dca466c03ce4e13b8398184ca610f015d69.png",
          description:
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
          socialLinks: {
            facebook: "#",
            linkedin: "#",
            twitter: "#",
          },
        },
        {
          id: 2,
          name: "Dr. Carlos Nguyen",
          position: "Lead Game Designer",
          image:
            "https://i.postimg.cc/t4GY0K9k/9f0f3cb05cfe27298de8750b0852c4baea79a752.png",
          description:
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
          socialLinks: {
            facebook: "#",
            linkedin: "#",
            twitter: "#",
          },
        },
        {
          id: 3,
          name: "David Lee",
          position: "Data Scientist",
          image:
            "https://i.postimg.cc/t4GY0K9k/9f0f3cb05cfe27298de8750b0852c4baea79a752.png",
          description:
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
          socialLinks: {
            facebook: "#",
            linkedin: "#",
            twitter: "#",
          },
        },
        {
          id: 4,
          name: "Alicia Brown",
          position: "Esports & Community Manager",
          image:
            "https://i.postimg.cc/KvMRHGh4/73350dca466c03ce4e13b8398184ca610f015d69.png",
          description:
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
          socialLinks: {
            facebook: "#",
            linkedin: "#",
            twitter: "#",
          },
        },
      ],
      teamDescription: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc luctus erat velit, non interdum lorem efficitur eu. Etiam hendrerit tortor at libero auctor hendrerit. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc luctus erat velit, non interdum lorem efficitur eu. Etiam hendrerit tortor at libero auctor hendrerit.",
    };

    return mockData;
  }
}

export const investmentAPI = InvestmentAPI.getInstance();
