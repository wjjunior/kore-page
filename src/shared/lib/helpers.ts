import type { Ref } from "vue";

export const scrollToSection = (
  href: string,
  sectionMap: Record<string, Ref<HTMLElement | null>>,
  headerOffset: number = 120
) => {
  const sectionRef = sectionMap[href];

  if (sectionRef?.value) {
    const offsetTop = sectionRef.value.offsetTop - headerOffset;

    window.scrollTo({
      top: offsetTop,
      behavior: "smooth",
    });
  }
};

interface ErrorWithMessage {
  message: unknown;
}

function isErrorWithMessage(error: unknown): error is ErrorWithMessage {
  return typeof error === "object" && error !== null && "message" in error;
}

export const extractErrorMessage = (err: unknown): string | null => {
  if (typeof err === "string") return err;
  if (isErrorWithMessage(err)) {
    if (typeof err.message === "string") return err.message;
    return "An error occurred";
  }
  return null;
};
