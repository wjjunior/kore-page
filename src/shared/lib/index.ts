export * from "./navigation";
export * from "./types";
