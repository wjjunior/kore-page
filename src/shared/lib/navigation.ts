export interface NavigationItem {
  id: string;
  label: string;
  href: string;
  isActive?: boolean;
}

export const NavItems = {
  INVEST: "invest",
  FAQ: "faq",
  CONTACT: "contact",
  EDUCATION: "education",
} as const;

export type NavItemType = (typeof NavItems)[keyof typeof NavItems];
