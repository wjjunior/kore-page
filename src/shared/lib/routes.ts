export const ROUTES = {
  HOME: "/",
  INVEST: "/invest",
  FAQ: "/faq",
  CONTACT: "/contact",
  EDUCATION: "/education",
} as const;

export type RouteKey = keyof typeof ROUTES;
