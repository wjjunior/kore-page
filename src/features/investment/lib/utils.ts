import { useInvestmentStore } from "@/app";
import type { InvestmentBannerData } from "@/shared/lib/types";
import { computed, watch } from "vue";
import { useInvestmentDataSSR, useInvestmentData } from "../composables";
import { extractErrorMessage } from "@/shared/lib/helpers";

export function getDefaultInvestmentData(): InvestmentBannerData {
  return {
    daysLeft: 0,
    totalInvestors: 0,
    fundingGoal: 0,
    fundsRaised: 0,
    minimumInvestment: 0,
    deadline: "",
    typeOfSecurity: "",
    revenueShareDuration: 0,
    categories: [],
    companyName: "",
    companyDescription: "",
    website: "",
    logoSrc: "",
    thumbnailSrc: "",
    offeringTerms: [],
    documents: [],
    teamMembers: [],
    teamDescription: "",
    marketingPlan: "",
    faqItems: [],
  };
}

export function mapClientInvestmentData(
  getters: Record<string, any>
): InvestmentBannerData {
  return {
    daysLeft: getters.getDaysLeft.value,
    totalInvestors: getters.getTotalInvestors.value,
    fundingGoal: getters.getFundingGoal.value,
    fundsRaised: getters.getFundsRaised.value,
    minimumInvestment: getters.getMinimumInvestment.value,
    deadline: getters.getDeadline.value,
    typeOfSecurity: getters.getTypeOfSecurity.value,
    revenueShareDuration: getters.getRevenueShareDuration.value,
    categories: getters.getCategories.value,
    companyName: getters.getCompanyName.value,
    companyDescription: getters.getCompanyDescription.value,
    website: getters.getWebsite.value,
    logoSrc: getters.getLogoSrc.value,
    thumbnailSrc: getters.getThumbnailSrc.value,
    offeringTerms: getters.getOfferingTerms.value,
    documents: getters.getDocuments.value,
    teamMembers: getters.getTeamMembers.value,
    teamDescription: getters.getTeamDescription.value,
    marketingPlan: getters.getMarketingPlan.value,
    faqItems: getters.getFaqItems.value,
  };
}

export function useInvestmentPageController() {
  // SSR source
  const {
    loading: ssrLoading,
    error: ssrError,
    isDataReady: ssrIsDataReady,
    refresh: ssrRefresh,
    investmentInfo: ssrInvestmentInfo,
  } = useInvestmentDataSSR();

  // Client fallback
  const {
    loading: clientLoading,
    error: clientError,
    isDataReady: clientIsDataReady,
    loadData: clientLoadData,
    retryLoad: clientRetryLoad,
    ...clientGetters
  } = useInvestmentData();

  const useClientData = computed(() => ssrError.value !== null);

  const loading = computed(() =>
    useClientData.value ? clientLoading.value : ssrLoading.value
  );

  const error = computed(() =>
    useClientData.value ? clientError.value : ssrError.value
  );

  const isDataReady = computed(() =>
    useClientData.value ? clientIsDataReady.value : ssrIsDataReady.value
  );

  const investmentData = computed<InvestmentBannerData>(() => {
    if (useClientData.value) {
      return mapClientInvestmentData(clientGetters);
    }
    return ssrInvestmentInfo.value ?? getDefaultInvestmentData();
  });

  const errorMessage = computed(() => extractErrorMessage(error.value));

  const refresh = () => {
    if (useClientData.value) {
      clientRetryLoad();
    } else {
      ssrRefresh();
    }
  };

  const investmentStore = useInvestmentStore();
  watch(
    () => ssrIsDataReady.value,
    (ready) => {
      if (ready && !ssrError.value && !investmentStore.isDataLoaded) {
        investmentStore.bannerData = ssrInvestmentInfo.value!;
      }
    },
    { immediate: true }
  );

  // Trigger client fallback when SSR fails
  watch(
    () => ssrError.value,
    (err) => {
      if (err !== null) {
        clientLoadData();
      }
    },
    { immediate: false }
  );

  return {
    loading,
    errorMessage,
    isDataReady,
    investmentData,
    refresh,
  };
}
