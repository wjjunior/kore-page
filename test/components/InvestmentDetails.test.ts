import { describe, it, expect, vi, beforeEach } from "vitest";
import { mount, flushPromises } from "@vue/test-utils";
import InvestmentDetails from "@/features/investment/ui/InvestmentDetails.vue";
import * as scrollUtils from "@/shared/lib";
import { mockInvestmentBannerData, createInvestmentStubs } from "../mocks";

const mountInvestmentDetails = async (props = {}) => {
  const wrapper = mount(InvestmentDetails, {
    props: {
      investmentData: mockInvestmentBannerData,
      loading: false,
      ...props,
    },
    global: {
      stubs: createInvestmentStubs(),
    },
  });
  await flushPromises();
  return wrapper;
};

describe("InvestmentDetails", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("should render the component", async () => {
    const wrapper = await mountInvestmentDetails();
    expect(wrapper.exists()).toBe(true);
  });

  it("should render all key sections", async () => {
    const wrapper = await mountInvestmentDetails();
    expect(wrapper.find(".stub-investment-cards").exists()).toBe(true);
    expect(wrapper.find(".stub-investment-team").exists()).toBe(true);
    expect(wrapper.find(".stub-investment-marketing").exists()).toBe(true);
    expect(wrapper.find(".stub-investment-faq").exists()).toBe(true);
  });

  it("should display navigation links", async () => {
    const wrapper = await mountInvestmentDetails();
    const navButtons = wrapper.findAll("nav button");
    const navTexts = navButtons.map((b) => b.text());
    expect(navTexts).toEqual([
      "Offering Terms and Documents",
      "Team",
      "Marketing Plan",
      "FAQ",
    ]);
  });

  it("should call scrollToSection on navigation click", async () => {
    const spy = vi.spyOn(scrollUtils, "scrollToSection");
    const wrapper = await mountInvestmentDetails();
    const firstButton = wrapper.find("nav button");
    await firstButton.trigger("click");
    expect(spy).toHaveBeenCalledWith("#offering-terms", expect.any(Object));
  });

  it("should pass loading state to InvestmentCards", async () => {
    const wrapper = await mountInvestmentDetails({ loading: false });
    const investmentCards = wrapper.findComponent({ name: "InvestmentCards" });
    expect(investmentCards.props("loading")).toBe(false);
  });

  it("should pass loading state to InvestmentCards when loading is true", async () => {
    const wrapper = await mountInvestmentDetails({ loading: true });
    const investmentCards = wrapper.findComponent({ name: "InvestmentCards" });
    expect(investmentCards.props("loading")).toBe(true);
  });

  it("should match snapshot", async () => {
    const wrapper = await mountInvestmentDetails();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
