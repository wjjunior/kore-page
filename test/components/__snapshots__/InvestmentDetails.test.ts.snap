// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InvestmentDetails > should match snapshot 1`] = `
"<div class="bg-white pt-6 pb-10 sm:pt-8 sm:pb-12 md:pt-10 md:pb-14 lg:pt-[93px] lg:pb-[137px]">
  <div class="max-w-7xl mx-auto px-4 sm:px-6 lg:px-8 xl:px-16">
    <div class="flex flex-col lg:flex-row gap-8 lg:gap-12">
      <div class="w-full lg:w-[280px] lg:flex-shrink-0">
        <div class="sticky top-8">
          <nav class="space-y-2"><button class="block text-left text-primary-200 font-hanken-grotesk font-bold text-base leading-5 hover:text-primary-300 transition-colors cursor-pointer">Offering Terms and Documents</button><button class="block text-left text-primary-200 font-hanken-grotesk font-bold text-base leading-5 hover:text-primary-300 transition-colors cursor-pointer">Team</button><button class="block text-left text-primary-200 font-hanken-grotesk font-bold text-base leading-5 hover:text-primary-300 transition-colors cursor-pointer">Marketing Plan</button><button class="block text-left text-primary-200 font-hanken-grotesk font-bold text-base leading-5 hover:text-primary-300 transition-colors cursor-pointer">FAQ</button></nav>
        </div>
      </div>
      <div class="w-full lg:w-[906px] lg:flex-shrink-0">
        <div id="offering-terms">
          <div class="stub-investment-cards">Offering Terms Documents</div>
        </div>
        <div id="team" class="mt-10">
          <div class="stub-investment-team">Team</div>
        </div>
        <div id="marketing-plan" class="mt-10">
          <div class="stub-investment-marketing">Marketing Plan</div>
        </div>
        <div id="faq" class="mt-10">
          <div class="stub-investment-faq">FAQ</div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
